src/proto/Elrond.proto

```diff
@@ -9,17 +9,68 @@ syntax = "proto3";
 package TW.Elrond.Proto;
 option java_package = "wallet.core.jni.proto";
 
-// A transaction, typical balance transfer
+// Generic / general-purpose transaction. Using one of the more specific messages (below) is recommended.
+// TODO: Ideally, we should rename this to "Transaction" or "GenericTransaction". The renaming is temporarily postponed (as of December 2021)
```

Comment:
There are many common fields in `EGLDTransfer`, `ESDTTransfer ` and `ESDTNFTTransfer` ( nonce/value/sender/receiver/gas/chain id), please take a look at `Cosmos.proto`, we can add a nested `Transaction` message to represent different transfer types

It's ok to do breaking changes as long as we have good test coverage (and mention it in the pr)

Reply:
Indeed.

Would the structure below be all right on your side (with breaking change on `SigningInput`)? 

Also, do you think it would be OK if we postpone this breaking change to the next PR? If not, let me know, and I will include the change in this PR.

```
SigningInput:
        bytes private_key
	string chain_id
	Gas gas

	oneof:
		- GenericTransaction
		- EGLDTransfer
		- ESDTTransfer
		- ESDTNFTTransfer

Gas:
    uint64      gas_limit
    uint64      gas_price

		
GenericTransaction:
	Accounts accounts
	string      value
	string      data
	uint32      version
	uint32      options

EGLDTransfer:
	Accounts accounts
	string      amount

ESDTTransfer:
	Accounts accounts
	string      amount
	string      token_identifier

ESDTNFTTransfer:
	Accounts accounts
	string      amount
	string      token_collection
    	uint64      token_nonce

Accounts:
    uint64    sender_nonce
    string      receiver
    string      sender
    string      sender_username
    string      receiver_username
```

Reply:
LGTM, no need to add `Gas` and just use `gas_limit` and `gas_price`

Reply:
I've updated the `Elrond.proto` file and made the necessary changes - mostly in `TransactionFactory.cpp` and in the test files. A new class has been defined, as well: `Transaction`. 

Perhaps it would make sense to merge the file `Serialization.cpp` into the new `Transaction.cpp`, in a future PR. Another (unrelated) idea for a future PR would be to move `Transaction` construction logic (currently in `TransactionFactory`) and gas estimation logic (currently in `GasEstimator`) into subclasses of the (abstract) `Transaction` class, such as: `GenericActionTransaction`, `EGLDTransferTransaction`, `ESDTTransferTransaction` etc.